tests/qml/tst_VPNTextArea.qml

```diff
@@ -0,0 +1,88 @@
+import QtQuick 2.3
+import QtTest 1.0
+
+import components.forms 0.1
+
+Item {
+    width: 600;
+    height: 800;
+
+    VPNTextArea {
+        id: vpnTextAreaTest
+    }
+
+    TestCase {
+        name: "VPNTextArea"
+        when: windowShown
+
+        function resetTextArea() {
+            vpnTextAreaTest.textInput.text = "";
+            vpnTextAreaTest.textInput.forceActiveFocus();
+        }
+
+        function insertSingleChar(character) {
+            keyClick(character);
+        }
+
+        function insertMultipleChars(chars) {
+            [...chars].forEach(character => {
+                insertSingleChar(character);
+            });
+        }
+
+        function fillTextInput() {
+            const maxCharLength = vpnTextAreaTest.textInput.maxCharacterCount;
+            const initialString = "a".repeat(maxCharLength);
+            vpnTextAreaTest.textInput.text = initialString;
+        }
+
+        // Insert single char within limit
+        function test_insertSingleChar() {
+            resetTextArea();
+
+            const expected = "a";
+            insertSingleChar(expected);
+
+            const actual = vpnTextAreaTest.textInput.text;
+            verify(expected === actual, `text was ${actual} not ${expected}.`);
+        }
+
+        // Insert multiple chars within limit
+        function test_insertMultipleChars() {
+            resetTextArea();
+
+            const expected = "abc";
+            insertMultipleChars(expected);
+
+            const actual = vpnTextAreaTest.textInput.text;
+            verify(expected === actual, `text was ${actual} not ${expected}.`);
+        }
+
+        // Insert single char exceeding limit
+        function test_exceedsCharLimitSingle() {
+            resetTextArea();
+            fillTextInput();
+
+            const expected = vpnTextAreaTest.textInput.text;
+            const inputString = "b";
+            insertSingleChar(inputString);
+
+            const actual = vpnTextAreaTest.textInput.text;
+            verify(expected === actual, `text was ${actual} not ${expected}.`);
+        }
+
+        // Insert multiple char exceeding limit
+        function test_exceedsCharLimitMultiple() {
+            resetTextArea();
+            fillTextInput();
+
+            const expected = vpnTextAreaTest.textInput.text;
+            const inputString = "abc";
+            insertMultipleChars(inputString);
+
+            const actual = vpnTextAreaTest.textInput.text;
+            verify(expected === actual, `text was ${actual} not ${expected}.`);
```

Comment:
When I first read this, I expected it to pass, but for the wrong reason. I thought that you'd copied by reference and so actual was a reference to expected. I see that this isn't the case and the test works correctly.

BUT I still think this would be clearer as 

const expected = "a".repeat(1000)

Where you're being explicit about the expected outcome and the actual is what you read off the property.

Reply:
You are right. It’s very easy to misunderstand what is happening here at first glance.